app.ts

```diff
@@ -0,0 +1,205 @@
+/** @format */
+
+interface IUser {
+    firstName: string;
+    secondName: string;
+    age: number;
+    gender: string;
+    isAdmin?: boolean;
+    specialization?: string;
+    course?: number;
+    faculty?: string;
+}
+
+class DataBase implements DataAggregator {
+    private static instance: DataBase;
+    private users: IUser[] = [];
+
+    public static getInstance(): DataBase {
+        if (!DataBase.instance) {
+            DataBase.instance = new DataBase();
+        }
+
+        return DataBase.instance;
+    }
+
+    public addUser(user: IUser): void {
+        this.users.push(user);
+    }
+
+    public getUsers(): IUser[] {
+        return this.users;
+    }
+
+    public removeUser(value: string, key: string): IUser[] {
+        return this.users = this.users.filter((item: IUser) => item[key] !== value)
+    }
+
+    public createIterator(): DataBaseIterator {
+        return new DataBaseIterator(this)
+    }
+}
+
+interface DataAggregator {
+    getUsers(): IUser[];
+}
+
+class DataBaseIterator {
+    private position: number = 0;
+
+    constructor(private users: DataAggregator) { }
+
+    public current(index: number = this.position) {
+        this.position = index;
+        return this.users.getUsers()[index];
+    }
+
+    public getUsers() {
+        return this.users.getUsers()
+    }
+
+    public next(): void {
+        this.position++;
+    }
+
+    public prev(): void {
+        this.position--;
+    }
+}
+
+class UserBuilder {
+
+    user: IUser;
+
+    constructor(role: string) {
+        this.role = role
```

Comment:
We have already discussed that Builder classes shouldn't hold any data, why are you assigning data to the builder instance again? Please carefully read the design of the Builder Pattern
https://metanit.com/sharp/patterns/2.5.php
The data should be assigned directly to the user class, which is created inside the Builder